### DISCUSSION TITLE
MTC output over usb not working as expected

### DISCUSSION THREAD

**Discussion by edwar64896** [FortySevenEffects/arduino_midi_library#289] [General]
I'm building an LTC (Linear TimeCode) to MTC (Midi TimeCode) converter with ATMEGA32u4 off an arduino pro micro board. I have had this working before on a Mega2560 with LUFA but want to condense the hardware, hence using native usb. LTC and decoding reading is working fine. MTC output is not.

I'm sending quarter frames at the rate of 4 per frame, as per MIDI spec and running at 25fps, however I'm only seeing one, maybe two messages per second coming into my midi monitor. I've tried a couple of midi-monitors and the result is the same. 

I've got a digital pin on toggle every time the MIDI.sendTimeCodeQuarterFrame(toSend) function is invoked, so I know I'm getting enough activity into that function, it's just not appearing at the output as regularly as necessary. 

Any ideas as to what might be happening here?

**Comment by franky47** [comment]
cc @lathoub, could this be a flush issue in the USB transport?

**Comment by lathoub** [comment]
> could this be a flush issue in the USB transport?

Unlikely, but not ruling it out.

@edwar64896 any chance you can `println` the content of the packet at
https://github.com/lathoub/Arduino-USBMIDI/blob/ab62611b9ef50a32875910d4af74eea260bbdcde/src/USB-MIDI_defs.h#L49

packet def:
https://github.com/arduino-libraries/MIDIUSB/blob/578f9f535e37cf8f557bddc1d6f85540b42cfba8/src/MIDIUSB_Defs.h#L4-L10

**Reply by edwar64896** [reply]
`

      byte midiMsg[2];

      /* send MIDI Timecode Quarter Frame*/
      void sendMTC() {

      digitalWrite(7,pin7); // debugging only
      pin7=1-pin7;
      
      switch (fps) {
        case 24:
          fps_snd = 0x70;
          break;
        case 25:
          fps_snd = 0x72;
          break;
        case 30:
          if (drop_frame_flag)
            fps_snd = 0x74;
          else
            fps_snd = 0x76;
          break;
        default:
          fps_snd = 0x72;
          break;
        }

    switch (MTCIndex)
    {
      case 0:
        toSend = ( 0x00 + (LTCIn_offset.f & 0xF));
        break;

      case 1:
        toSend = ( 0x10 + ((LTCIn_offset.f & 0xF0) / 16));
        break;

      case 2:
        toSend = ( 0x20 + (LTCIn_offset.s & 0xF));
        break;

      case 3:
        toSend = ( 0x30 + ((LTCIn_offset.s & 0xF0) / 16));
        break;

      case 4:
        toSend = ( 0x40 + (LTCIn_offset.m & 0xF));
        break;

      case 5:
        toSend = ( 0x50 + ((LTCIn_offset.m & 0xF0) / 16));
        break;

      case 6:
        toSend = ( 0x60 + (LTCIn_offset.h & 0xF));
      break;

      case 7:
        toSend = ( fps_snd + ((LTCIn_offset.h & 0xF0) / 16)); // 0x70 = 24 fps // 0x72 = 25 fps // 0x74 = 30df fps // 0x76 = 30 fps
        break;
    }

    //MIDI.sendTimeCodeQuarterFrame(toSend); 

    midiMsg[0]=0xf1;
    midiMsg[1]=toSend;
    MIDI.sendSysEx(sizeof(midiMsg),midiMsg,true);


    if (++MTCIndex > 7)
      MTCIndex = 0;
      }

`

I'll try and organise that for you - in the meantime, code supplied above. 

I have tried both MIDI.sendSysEx and also MIDI.sendTimeCodeQuarterFrame options. Obvs prefer the latter and had this working with mocolufa and the Mega2560 after some lufa patches, but I think we discussed that back in 2018.

I'm getting 4 toggles on the debug pin for every LTC frame so I know this function is getting hit.

**Comment by edwar64896** [comment]
![Screenshot 2022-07-06 204459](https://user-images.githubusercontent.com/2887548/177533727-cd645076-d5d7-44dc-bae6-dfa8d4e46b5e.png)
![Screenshot 2022-07-06 204442](https://user-images.githubusercontent.com/2887548/177533734-d8886eb4-2233-4e64-a691-f53e986c5924.png)

ok so interesting outcome from testing. The first image is my midi monitor, showing a paucity of messages appearing on the midi input, however the second is my println output at sendMidi. Lots of messages here, so sendMidi is definitely getting called with adequate data.

What is _spooky_ is that by reviewing both these outputs concurrently, we notice that:

1. approximately every 5th message being sent actually makes it out as a midi message
2. only the 2-byte (>0x0F) messages seem to be stable. where a message byte that is lower than 0x0F makes it through, sometimes it doesn't match what is in the println buffer. For instance, 11th line from the bottom in the coolterm printout corresponds with line 4 from the bottom on the midi monitor output, however the bytes are not the same. 0x0F on coolterm as opposed to 0x05 on the midi monitor. This maketh no sense to me. Surrounding message bytes that got through do correspond to the coolterm messages.

**Reply by edwar64896** [reply]
@lathoub @franky47 Do you think it would be necessary at this point to create an issue relating to this behaviour?

**Reply by lathoub** [reply]
maybe, not sure at this stage. And if so, where to create the issue?: here, [Arduino-USBMIDI Transport](https://github.com/lathoub/Arduino-USBMIDI) or [MIDIUSB](https://github.com/arduino-libraries/MIDIUSB)

Can you create a minimum sketch for me to work from?

**Reply by edwar64896** [reply]
Hey. I am unfortunately going to have to leave this in your capable hands for a little while. I am overseas in UK visiting family and will not have access to my development machine for a few weeks.
